ml/src/main/scala/org/apache/spark/ml/FrameMLUtils.scala

```diff
@@ -0,0 +1,132 @@
+                      h2oContext: H2OContext,
+                      sqlContext: SQLContext): (RDD[LabeledPoint], Array[Double]) = {
+    var means: Array[Double] = new Array[Double](nfeatures)
+    val domains = frame.domains()
+
+    val trainingDF = h2oContext.asDataFrame(new H2OFrame(frame))(sqlContext)
+    val fields: Array[StructField] = trainingDF.schema.fields
+    var trainingRDD = trainingDF.rdd
+
+    if (MissingValuesHandling.Skip.eq(missingHandler)) {
+      trainingRDD = trainingRDD.filter(!_.anyNull)
+    } else if (MissingValuesHandling.MeanImputation.eq(missingHandler)) {
+      // Computing the means by hand and not using frame.means() as it does not compute the mean for enum columns
+      means = movingAverage(trainingRDD, fields, domains)
+    }
+
+    (trainingRDD.map(row => {
+      val features = new Array[Double](nfeatures)
+      (0 until nfeatures).foreach(i => features(i) = if (row.isNullAt(i)) means(i) else toDouble(row.get(i), fields(i), domains(i)))
+
+      new LabeledPoint(
+        toDouble(row.getAs[String](_response_column), fields(fields.length - 1), domains(domains.length - 1)),
+        Vectors.dense(features)
+      )
+    }), means)
+  }
+
+  // Running average so we don't get overflows
+  private[ml] def movingAverage(trainingRDD: RDD[Row],
+                                fields: Array[StructField],
+                                domains: Array[Array[String]]): Array[Double] = {
+    val means = new Array[Double](fields.length)
+    val counts = new Array[Int](means.length)
+    trainingRDD.aggregate(means.zip(counts))(
+      // Compute the average within a RDD partition
+      (agg, row) => {
+        agg.indices.foreach(i => {
+          if (!row.isNullAt(i)) {
+            val value = toDouble(row.get(i), fields(i), domains(i))
+            val delta = value - agg(i)._1
+            val n = agg(i)._2 + 1
+            agg(i) = (agg(i)._1 + delta / n, n)
+          }
+        })
+        agg
+      },
+      // Merge all RDD partition stats
+      (agg1, agg2) => {
+        merge(agg1, agg2)
+      }
+    ).map(_._1)
+  }
+
+  def merge(agg1: Array[(Double, Int)], agg2: Array[(Double, Int)]): Array[(Double, Int)] = {
+    agg1.indices.foreach { idx =>
+      if (agg1(idx)._2 == 0) {
+        agg1(idx) = agg2(idx)
+      } else {
+        val otherMu: Double = agg2(idx)._1
+        val mu: Double = agg1(idx)._1
+        val n = agg1(idx)._2
+        val otherN = agg2(idx)._2
+        val delta = otherMu - mu
+
+        if (otherN * 10 < n) {
+          agg1(idx) = (mu + (delta * otherN) / (n + otherN), n + otherN)
+        } else if (n * 10 < otherN) {
+          agg1(idx) = (otherMu - (delta * n) / (n + otherN), n + otherN)
+        } else {
+          agg1(idx) = ((mu * n + otherMu * otherN) / (n + otherN), n + otherN)
+        }
+      }
+    }
+    agg1
+  }
+
+  private[ml] def toDouble(value: Any, fieldStruct: StructField, domain: Array[String]): Double = {
```

Comment:
This is not very type safe; why not check value for its type, in match/case?

E.g. value match {
  case b; Byte if fieldStruct.dataType == DataTypes.ByteType => b.doubleValue
... etc ...

And I would make it less verbose.